doc/developer/design/20220516_transactional_consistency.md

```diff
@@ -0,0 +1,416 @@
+- Transaction T1 is ordered before Transaction T2 if the timestamp of T1 == the timestamp of T2, T1 is write-only, and
+  T2 is read-only.
+- Transaction T1 can be arbitrarily ordered with Transaction T2 if the timestamp of T1 == the timestamp of T2, T1 is
+  write-only, and T2 is write-only.
+- Transaction T1 can be arbitrarily ordered with Transaction T2 if the timestamp of T1 == the timestamp of T2, T1 is
+  read-only, and T2 is read-only.
+
+#### Strict Serializable
+
+See [https://jepsen.io/consistency/models/strict-serializable](https://jepsen.io/consistency/models/strict-serializable)
+for an in depth formal discussion of Strict Serializable. Strict Serializable is a combination of Serializable and
+Linearizable, which means that the total order formed by all transactions must be constrained by their real time
+occurrences. So if transaction t1 happened in real time before transaction t2, then t1 must be ordered before t2 in the
+total ordering.
+
+Operations that only include user tables are currently Strict Serializable. This is made easy by using a single global
+timestamp and a single thread in the coordinator to order operations. Writes happen at a timestamp larger than all
+previous operations. Reads happen at a timestamp equal to or larger than all previous operations.
+
+Operations that involve reading from views over upstream data are not currently Strict Serializable. It is possible to
+read from a view at one timestamp and then later read from the same view at an earlier timestamp. Consider the following
+scenario:
+
+1. Client reads from view v1 that has an upper of `ts + 1` and a since of `ts - 10`. The read will happen at `ts`.
+2. Client reads from a join over v1 and view v2. v2 has an upper of `ts - 5` and some since less or equal to `ts - 6`.
+   The read will happen at `ts - 6`.
+
+Reference [Timestamp Selection](#Timestamp Selection) to see why these timestamps are selected. This is an obvious
+violation of strict serializability because the second read happens later than the first read in real time, but it is
+assigned an earlier timestamp.
+
+In order to achieve Strict Serializability Materialize must always select a timestamp greater than all previously
+selected timestamps. This will ensure that if an operation happens later in real time, then it will be assigned a later
+timestamp. Consecutive reads can all be assigned an equal timestamp as long as there’s no writes in between them. This
+is made somewhat easy by the fact that all timestamp selection happens on a single thread so no coordination is needed.
+On implementation for this is to consult the global timestamp for all reads, not just ones involving user tables. If
+the `[since, upper)` of any object drifts ahead of the global timestamp, causing the selected timestamp to be greater
+than the global timestamp, then the global timestamp can be fast-forwarded to the selected timestamp.
```

Comment:
Yes I was referring to https://dev.materialize.com/api/rust-private/mz_coord/coord/timeline/struct.TimestampOracle.html#method.fast_forward. Oy, the more I think about my proposed approach the more issues I'm seeing with it.

I just want to lay out an example that demonstrates the issue you're talking about:
1. Some view v has an upper of `ts + 1000` and a since of `ts + 999`.
2. All other views have an upper of `ts + 1` a since of `ts`.
3. The global timestamp is `ts`.
4. The client reads from v, the selected timestamp will be `ts + 999`, which fast forwards the global timestamp to `ts + 999`.
5. The client tries to read from another view, the selected timestamp will be `ts + 999` again, which blocks the read until the view catches up. 

IF all views had roughly the same `upper` and `since` I don't think this would cause a problem, but I don't know if we can rely on that assumption. I was (probably wrongly) assuming that all `upper`s were roughly equal to the current wall clock. I need to get a better understanding of what determines the [`since`, `upper`) ranges for real-time sources.

However, I think there's an opposite (but smaller) issue with your proposed solution of "have the query wait instead of it forcing all other queries to wait". The scenario would look like the following:
1. Some view v has an upper of `ts + 1000` and a since of `ts + 999`.
2. All other views have an upper of `ts + 1` a since of `ts`.
3. The global timestamp is `ts`.
4. The client reads from v, the selected timestamp will be `ts + 999`, which will cause the query to block until the global timestamp catches up to `ts + 999`. 
5. If the client tries to read from another view, the selected timestamp will be `ts`, which can return immediately.

The issue here is that every time the client tries to read from v, the read will be blocked, which is still a pretty bad UX if v is read often.

In general I think if the [`since`, `upper`) range of two objects diverge, it becomes very difficult to order operations between them because they have different notions of "now". So either something needs to block or we need to prevent the ranges from diverging.

I like your suggestion below about holding back `since`s, that could help ensure that there's some timestamp that's valid for all objects without blocking. The hybrid logical timestamps could actually also help this problem because it would help prevent real-time sources from advancing their `upper`s past the current wall clock, meaning all `upper`s would be within 1 ms of each other and no query would need to block for more than 1 ms.

Reply:
There's also definitely a connection between the solutions that fix the problem you pointed out here and the solutions that fix the problem of the global timestamp advancing past the wall clock (discussed below in the doc). I think I need to update this doc tomorrow that makes the connection more clear.

Reply:
Ok, just pushed some updates that I think make this a bit more unified and clearer. Let me know what you think.